bpfd/src/bpf.rs

```diff
@@ -207,148 +135,205 @@ impl<'a> BpfManager<'a> {
-            let _ = fs::remove_dir_all(map_pin_path);
-            BpfdError::SectionNameNotValid(section_name.clone())
-        })?;
-
-        // Calculate the next_available_id
-        let next_available_id = if let Some(prog) = self.programs.get(&if_index) {
-            prog.len() + 1
-        } else {
-            self.programs.insert(if_index, HashMap::new());
-            1
-        };
-        if next_available_id > 10 {
-            return Err(BpfdError::TooManyPrograms);
+    pub(crate) fn add_program(&mut self, p: Program) -> Result<Uuid, BpfdError> {
+        match p {
+            Program::Xdp(_, _) => self.add_xdp_program(p),
+            Program::Tracepoint(_, _) => self.add_single_attach_program(p),
         }
+    }
 
-        // Calculate the dispatcher revision
-        let (old_revision, revision) = if let Some(old_revision) = self.revisions.remove(&if_index)
-        {
-            let r = old_revision.wrapping_add(1);
-            self.revisions.insert(if_index, r);
-            (Some(old_revision), r)
-        } else {
-            self.revisions.insert(if_index, 0);
-            (None, 0)
-        };
+    pub(crate) fn add_xdp_program(&mut self, p: Program) -> Result<Uuid, BpfdError> {
+        if let Program::Xdp(ref data, ref info) = p {
+            let if_index = info.if_index;
+            let id = Uuid::new_v4();
+            let map_pin_path = format!("/var/run/bpfd/fs/maps/{}", id);
+            fs::create_dir_all(map_pin_path.clone())?;
+
+            let mut ext_loader = BpfLoader::new()
+                .extension(&data.section_name)
+                .map_pin_path(map_pin_path.clone())
+                .load_file(&data.path)?;
+
+            ext_loader.program_mut(&data.section_name).ok_or_else(|| {
+                let _ = fs::remove_dir_all(map_pin_path);
+                BpfdError::SectionNameNotValid(data.section_name.clone())
+            })?;
+
+            // Calculate the next_available_id
+            let next_available_id = self
+                .programs
+                .iter()
+                .filter(|(_, p)| {
+                    if let Program::Xdp(_, i) = p {
+                        i.if_index == info.if_index
+                    } else {
+                        false
+                    }
+                })
+                .collect::<HashMap<_, _>>()
+                .len();
+            if next_available_id > 10 {
+                return Err(BpfdError::TooManyPrograms);
+            }
 
-        let prog = ExtensionProgram {
-            path,
-            current_position: None,
-            metadata: Metadata {
-                priority,
-                name: section_name,
-                attached: false,
-            },
-            owner,
-            proceed_on,
-        };
-        prog.save(id)
-            .map_err(|_| BpfdError::Error("unable to persist program data".to_string()))?;
+            // Calculate the dispatcher revision
```

Comment:
A question I've had is why we need to keep the dispatcher revision.  I may be missing it, but I'm not seeing where you're using it for anything.

Reply:
The revision is used to allow us to keep 2 sets of extension programs attached at the same time to facilitate atomic replacement
i.e these paths can co-exist:
`dispatcher_1_1/prog_1_link`
`dispatcher_1_2/prog_1_link`

And when we've atomically replaced the dispatcher we can delete `dispatcher_1_1`